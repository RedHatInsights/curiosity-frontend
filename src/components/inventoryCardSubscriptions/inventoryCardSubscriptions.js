import React from 'react';
import {
  useGetSubscriptionsInventory,
  useInventoryCardActionsSubscriptions,
  useOnPageSubscriptions,
  useOnColumnSortSubscriptions,
  useParseSubscriptionsFiltersSettings
} from './inventoryCardSubscriptionsContext';
import { InventoryCard } from '../inventoryCard/inventoryCard';
import { helpers } from '../../common';

/**
 * @memberof Components
 * @module InventoryCardSubscriptions
 * @property {module} InventoryCardSubscriptionsContext
 */

/**
 * A subscriptions' system inventory component.
 *
 * @param {object} props
 * @param {boolean} [props.isDisabled=helpers.UI_DISABLED_TABLE_SUBSCRIPTIONS]
 * @param {useGetSubscriptionsInventory} [props.useGetInventory=useGetSubscriptionsInventory]
 * @param {useInventoryCardActionsSubscriptions} [props.useInventoryCardActions=useInventoryCardActionsSubscriptions]
 * @param {useOnPageSubscriptions} [props.useOnPage=useOnPageSubscriptions]
 * @param {useOnColumnSortSubscriptions} [props.useOnColumnSort=useOnColumnSortSubscriptions]
 * @param {useParseSubscriptionsFiltersSettings} [props.useParseFiltersSettings=useParseSubscriptionsFiltersSettings]
 * @returns {JSX.Element}
 */
const InventoryCardSubscriptions = ({
  isDisabled = helpers.UI_DISABLED_TABLE_SUBSCRIPTIONS,
  useGetInventory = useGetSubscriptionsInventory,
  useInventoryCardActions = useInventoryCardActionsSubscriptions,
  useOnPage = useOnPageSubscriptions,
  useOnColumnSort = useOnColumnSortSubscriptions,
  useParseFiltersSettings = useParseSubscriptionsFiltersSettings
}) => (
  <InventoryCard
    isDisabled={isDisabled}
    useGetInventory={useGetInventory}
    useInventoryCardActions={useInventoryCardActions}
    useOnPage={useOnPage}
    useOnColumnSort={useOnColumnSort}
    useParseFiltersSettings={useParseFiltersSettings}
  />
);

export { InventoryCardSubscriptions as default, InventoryCardSubscriptions };
