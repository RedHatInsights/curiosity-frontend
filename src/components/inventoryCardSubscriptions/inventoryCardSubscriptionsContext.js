import { reduxActions, reduxTypes, storeHooks } from '../../redux';
import {
  useProduct,
  useProductInventorySubscriptionsConfig,
  useProductInventorySubscriptionsQuery
} from '../productView/productViewContext';
import {
  RHSM_API_QUERY_INVENTORY_SORT_DIRECTION_TYPES as SORT_DIRECTION_TYPES,
  RHSM_API_QUERY_INVENTORY_SUBSCRIPTIONS_SORT_TYPES as SORT_TYPES,
  RHSM_API_QUERY_SET_TYPES
} from '../../services/rhsm/rhsmConstants';
import { helpers } from '../../common';
import {
  useGetInstancesInventory,
  useInventoryCardActionsInstances,
  useParseInstancesFiltersSettings,
  useSelectorInstances
} from '../inventoryCardInstances/inventoryCardInstancesContext';
import { tableHelpers } from '../table/table';

/**
 * @memberof InventoryCardSubscriptions
 * @module InventoryCardSubscriptionsContext
 */

/**
 * Parse filters settings for context.
 * See @module InventoryCardInstancesContext
 *
 * @param {object} options
 * @param {boolean} options.isDisabled
 * @param {Function} options.useParseFiltersSettings
 * @param {Function} options.useProductConfig
 * @returns {{settings: {}, columnCountAndWidths: {count: number, widths: Array}, filters: Array}}
 */
const useParseSubscriptionsFiltersSettings = ({
  isDisabled = false,
  useParseFiltersSettings: useAliasParseFiltersSettings = useParseInstancesFiltersSettings,
  useProductConfig: useAliasProductConfig = useProductInventorySubscriptionsConfig
} = {}) =>
  useAliasParseFiltersSettings({
    isDisabled,
    useProductConfig: useAliasProductConfig
  });

/**
 * Parse selector response for consuming components.
 * See @module InventoryCardInstancesContext
 *
 * @param {object} options
 * @param {string} options.storeRef
 * @param {Function} options.useParseFiltersSettings
 * @param {Function} options.useProductInventoryQuery
 * @param {Function} options.useSelector
 * @returns {{pending: boolean, fulfilled: boolean, error: boolean, resultsColumnCountAndWidths: {count: number,
 *     widths: Array}, dataSetColumnHeaders: Array, resultsPerPage: number, resultsOffset: number, dataSetRows: Array,
 *     resultsCount: number}}
 */
const useSelectorSubscriptions = ({
  storeRef = 'subscriptionsInventory',
  useParseFiltersSettings: useAliasParseFiltersSettings = useParseSubscriptionsFiltersSettings,
  useProductInventoryQuery: useAliasProductInventoryQuery = useProductInventorySubscriptionsQuery,
  useSelector: useAliasSelector = useSelectorInstances
} = {}) =>
  useAliasSelector({
    storeRef,
    useParseFiltersSettings: useAliasParseFiltersSettings,
    useProductInventoryQuery: useAliasProductInventoryQuery
  });

/**
 * Combine service call, Redux, and inventory selector response.
 * See @module InventoryCardInstancesContext
 *
 * @param {object} options
 * @param {boolean} options.isDisabled
 * @param {Function} options.getInventory
 * @param {Function} options.useGetInventory
 * @param {Function} options.useProductInventoryQuery
 * @param {Function} options.useSelector
 * @returns {{pending: boolean, fulfilled: boolean, error: boolean, resultsColumnCountAndWidths: {count: number,
 *     widths: Array}, dataSetColumnHeaders: Array, resultsPerPage: number, resultsOffset: number, dataSetRows: Array,
 *     resultsCount: number}}
 */
const useGetSubscriptionsInventory = ({
  isDisabled = false,
  getInventory = reduxActions.rhsm.getSubscriptionsInventory,
  useGetInventory: useAliasGetInventory = useGetInstancesInventory,
  useProductInventoryQuery: useAliasProductInventoryQuery = useProductInventorySubscriptionsQuery,
  useSelector: useAliasSelector = useSelectorSubscriptions
} = {}) =>
  useAliasGetInventory({
    isDisabled,
    getInventory,
    useProductInventoryQuery: useAliasProductInventoryQuery,
    useSelector: useAliasSelector
  });

/**
 * Return a component list for a configurable inventoryCard action toolbar.
 * Allow the "content" prop to receive inventory data for display via callback.
 * See @module InventoryCardInstancesContext
 *
 * @param {object} options
 * @param {Function} options.useInventoryCardActions
 * @param {Function} options.useSelector
 * @param {Function} options.useProductConfig
 * @returns {Array}
 */
const useInventoryCardActionsSubscriptions = ({
  useInventoryCardActions: useAliasInventoryCardActions = useInventoryCardActionsInstances,
  useSelector: useAliasSelector = useSelectorSubscriptions,
  useProductConfig: useAliasProductConfig = useProductInventorySubscriptionsConfig
} = {}) => useAliasInventoryCardActions({ useSelector: useAliasSelector, useProductConfig: useAliasProductConfig });

/**
 * An onPage callback for inventory.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useProduct
 * @returns {Function}
 */
const useOnPageSubscriptions = ({
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useProduct: useAliasProduct = useProduct
} = {}) => {
  const { productId } = useAliasProduct();
  const dispatch = useAliasDispatch();

  /**
   * On event update state for inventory.
   *
   * @event onPage
   * @param {object} params
   * @param {number} params.offset
   * @param {number} params.perPage
   * @returns {void}
   */
  return ({ offset, perPage }) => {
    dispatch([
      {
        type: reduxTypes.query.SET_QUERY_RHSM_SUBSCRIPTIONS_INVENTORY_TYPES[RHSM_API_QUERY_SET_TYPES.OFFSET],
        viewId: productId,
        [RHSM_API_QUERY_SET_TYPES.OFFSET]: offset
      },
      {
        type: reduxTypes.query.SET_QUERY_RHSM_SUBSCRIPTIONS_INVENTORY_TYPES[RHSM_API_QUERY_SET_TYPES.LIMIT],
        viewId: productId,
        [RHSM_API_QUERY_SET_TYPES.LIMIT]: perPage
      }
    ]);
  };
};

/**
 * An onColumnSort callback for inventory.
 *
 * @param {object} options
 * @param {object} options.sortColumns
 * @param {Function} options.useDispatch
 * @param {Function} options.useProduct
 * @returns {Function}
 */
const useOnColumnSortSubscriptions = ({
  sortColumns = SORT_TYPES,
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useProduct: useAliasProduct = useProduct
} = {}) => {
  const { productId } = useAliasProduct();
  const dispatch = useAliasDispatch();

  /**
   * On event update state for inventory.
   *
   * @event onColumnSort
   * @param {object} params
   * @param {string} params.direction
   * @param {object} params.data
   * @returns {void}
   */
  return ({ direction, data = {} }) => {
    const { metric: id } = data;
    const updatedSortColumn = Object.values(sortColumns).find(value => value === id);
    let updatedDirection;

    if (!updatedSortColumn) {
      if (helpers.DEV_MODE || helpers.REVIEW_MODE) {
        console.warn(`Sorting can only be performed on select fields, confirm field ${id} is allowed.`);
      }
      return;
    }

    switch (direction) {
      case tableHelpers.SortByDirectionVariant.desc:
        updatedDirection = SORT_DIRECTION_TYPES.DESCENDING;
        break;
      default:
        updatedDirection = SORT_DIRECTION_TYPES.ASCENDING;
        break;
    }

    dispatch([
      {
        type: reduxTypes.query.SET_QUERY_RHSM_SUBSCRIPTIONS_INVENTORY_TYPES[RHSM_API_QUERY_SET_TYPES.DIRECTION],
        viewId: productId,
        [RHSM_API_QUERY_SET_TYPES.DIRECTION]: updatedDirection
      },
      {
        type: reduxTypes.query.SET_QUERY_RHSM_SUBSCRIPTIONS_INVENTORY_TYPES[RHSM_API_QUERY_SET_TYPES.SORT],
        viewId: productId,
        [RHSM_API_QUERY_SET_TYPES.SORT]: updatedSortColumn
      }
    ]);
  };
};

const context = {
  useGetSubscriptionsInventory,
  useInventoryCardActionsSubscriptions,
  useOnPageSubscriptions,
  useOnColumnSortSubscriptions,
  useParseSubscriptionsFiltersSettings,
  useSelectorSubscriptions
};

export {
  context as default,
  context,
  useGetSubscriptionsInventory,
  useInventoryCardActionsSubscriptions,
  useOnPageSubscriptions,
  useOnColumnSortSubscriptions,
  useParseSubscriptionsFiltersSettings,
  useSelectorSubscriptions
};
