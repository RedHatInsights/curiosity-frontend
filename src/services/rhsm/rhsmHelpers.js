import { RHSM_API_QUERY_SET_TYPES } from './rhsmConstants';

/**
 * ToDo: remove filterArchitectureVariant when the API supports architecture, variant params
 */
/**
 * Patch for returning a made up API architecture, variant param as a product ID
 *
 * @param {string} id
 * @param {object} params
 * @returns {string}
 */
const filterArchitectureVariant = (id, params = {}) => {
  const updatedId = id;

  if (params?.[RHSM_API_QUERY_SET_TYPES.ARCHITECTURE]?.length) {
    return params?.[RHSM_API_QUERY_SET_TYPES.ARCHITECTURE];
  }

  if (params?.[RHSM_API_QUERY_SET_TYPES.VARIANT]?.length) {
    return params?.[RHSM_API_QUERY_SET_TYPES.VARIANT];
  }

  return updatedId;
};

const rhsmHelpers = {
  filterArchitectureVariant
};

export { rhsmHelpers as default, rhsmHelpers, filterArchitectureVariant };
