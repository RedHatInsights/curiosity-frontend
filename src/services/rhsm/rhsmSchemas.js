import JoiBase from 'joi';
import JoiDate from '@joi/date';
import { schemaResponse } from '../common/helpers';
import { rhsmConstants } from './rhsmConstants';

/**
 * @memberof Rhsm
 * @module RhsmSchemas
 */

/**
 * Extend Joi with date.
 *
 * @type {*}
 */
const Joi = JoiBase.extend(JoiDate);

/**
 * Error response item.
 *
 * @type {*} Joi schema
 */
const errorItem = Joi.object({
  code: Joi.string().default(null),
  detail: Joi.string().default(null)
}).unknown(true);

/**
 * Error response.
 *
 * @type {*} Joi schema
 */
const errorResponseSchema = Joi.object()
  .keys({
    errors: Joi.array().items(errorItem).default([])
  })
  .unknown(true);

const linksSchema = Joi.object();

/**
 * RHSM base response meta field.
 *
 * @type {*} Joi schema
 */
const metaResponseSchema = Joi.object()
  .keys({
    count: Joi.number().integer().default(0),
    product: Joi.string().valid(...Object.values(rhsmConstants.RHSM_API_PATH_PRODUCT_TYPES))
  })
  .unknown(true);

/**
 * Capacity response meta field.
 *
 * @type {*} Joi schema
 */
const capacityMetaSchema = metaResponseSchema
  .keys({
    metric_id: Joi.string().valid(...Object.values(rhsmConstants.RHSM_API_PATH_METRIC_TYPES))
  })
  .unknown(true);

/**
 * Capacity response item.
 *
 * @type {*} Joi schema
 */
const capacityItem = Joi.object({
  date: Joi.date().utc().allow(null),
  has_data: Joi.boolean().optional().allow(null),
  has_infinite_quantity: Joi.boolean().optional().allow(null),
  value: Joi.number().allow(null).default(0)
})
  .unknown(true)
  .default();

/**
 * Capacity response.
 *
 * @type {*} Joi schema
 */
const capacityResponseSchema = Joi.object().keys({
  data: Joi.array().items(capacityItem).default([]),
  links: linksSchema.default({}),
  meta: capacityMetaSchema.default({})
});

/**
 * Guests response meta field.
 *
 * @type {*} Joi schema
 */
const guestsMetaSchema = Joi.object()
  .keys({
    count: Joi.number().integer().default(0)
  })
  .unknown(true);

/**
 * Guests response item.
 *
 * @type {*} Joi schema
 */
const guestsItem = Joi.object({
  inventory_id: Joi.string().optional().allow(null),
  display_name: Joi.string().optional().allow(null),
  subscription_manager_id: Joi.string().optional().allow(null),
  last_seen: Joi.date().utc().allow(null)
})
  .unknown(true)
  .default();

/**
 * Guests response.
 *
 * @type {*} Joi schema
 */
const guestsResponseSchema = Joi.object().keys({
  data: Joi.array().items(guestsItem).default([]),
  links: linksSchema.default({}),
  meta: guestsMetaSchema.default({})
});

/**
 * Instances response meta field.
 *
 * @type {*} Joi schema
 */
const instancesMetaSchema = metaResponseSchema
  .keys({
    count: Joi.number().integer().default(0),
    measurements: Joi.array()
      .items(Joi.string().valid(...Object.values(rhsmConstants.RHSM_API_PATH_METRIC_TYPES)))
      .default([])
  })
  .unknown(true);

/**
 * Instances response item.
 *
 * @type {*} Joi schema
 */
const instancesItem = Joi.object({
  instance_id: Joi.string().optional().allow(null),
  category: Joi.string().lowercase().optional().allow(null),
  cloud_provider: Joi.string().lowercase().optional().allow(null, ''),
  display_name: Joi.string().optional().allow(null),
  billing_provider: Joi.string().lowercase().optional().allow(null, ''),
  billing_account_id: Joi.string().optional().allow(null),
  measurements: Joi.array().default([]),
  number_of_guests: Joi.number().integer().default(0),
  subscription_manager_id: Joi.string().optional().allow(null),
  last_seen: Joi.date().utc().allow(null)
})
  .unknown(true)
  .default();

/**
 * Instances response.
 *
 * @type {*} Joi schema
 */
const instancesResponseSchema = Joi.object().keys({
  data: Joi.array().items(instancesItem).default([]),
  links: linksSchema.default({}),
  meta: instancesMetaSchema.default({})
});

/**
 * Subscriptions response meta field.
 *
 * @type {*} Joi schema
 */
const subscriptionsMetaSchema = metaResponseSchema
  .keys({
    subscription_type: Joi.string().valid(null, ...Object.values(rhsmConstants.RHSM_API_RESPONSE_SUBSCRIPTION_TYPES))
  })
  .unknown(true);

/**
 * Subscriptions response item.
 *
 * @type {*} Joi schema
 */
const subscriptionsItem = Joi.object({
  has_infinite_quantity: Joi.boolean().optional().allow(null),
  next_event_date: Joi.date().utc().allow(null),
  product_name: Joi.string().optional().allow(null),
  quantity: Joi.number().allow(null).default(0),
  service_level: Joi.string().valid(...Object.values(rhsmConstants.RHSM_API_RESPONSE_SLA_TYPES)),
  total_capacity: Joi.number().allow(null).default(0),
  uom: Joi.string()
    .lowercase()
    .valid(...Object.values(rhsmConstants.RHSM_API_RESPONSE_UOM_TYPES))
})
  .unknown(true)
  .default();

/**
 * Subscriptions response.
 *
 * @type {*} Joi schema
 */
const subscriptionsResponseSchema = Joi.object().keys({
  data: Joi.array().items(subscriptionsItem).default([]),
  links: linksSchema.default({}),
  meta: subscriptionsMetaSchema.default({})
});

/**
 * Tally and capacity metric response item.
 *
 * @type {*} Joi schema
 */
const tallyItem = Joi.object({
  date: Joi.date().utc().allow(null),
  has_data: Joi.boolean().optional().allow(null),
  value: Joi.number().allow(null).default(0)
})
  .unknown(true)
  .default();

/**
 * Tally response meta field.
 *
 * @type {*} Joi schema
 */
const tallyMetaSchema = metaResponseSchema
  .keys({
    has_cloudigrade_data: Joi.boolean().optional().allow(null),
    has_cloudigrade_mismatch: Joi.boolean().optional().allow(null),
    metric_id: Joi.string().valid(...Object.values(rhsmConstants.RHSM_API_PATH_METRIC_TYPES)),
    total_monthly: tallyItem
  })
  .unknown(true);

/**
 * Tally response.
 *
 * @type {*} Joi schema
 */
const tallyResponseSchema = Joi.object().keys({
  data: Joi.array().items(tallyItem).default([]),
  links: linksSchema.default({}),
  meta: tallyMetaSchema.default({})
});

const rhsmSchemas = {
  capacity: response => schemaResponse({ response, schema: capacityResponseSchema, id: 'RHSM capacity' }),
  errors: response => schemaResponse({ response, schema: errorResponseSchema, id: 'RHSM errors' }),
  guests: response => schemaResponse({ response, schema: guestsResponseSchema, id: 'RHSM guests' }),
  instances: response => schemaResponse({ response, schema: instancesResponseSchema, id: 'RHSM instances' }),
  subscriptions: response =>
    schemaResponse({ response, schema: subscriptionsResponseSchema, id: 'RHSM subscriptions' }),
  tally: response => schemaResponse({ response, schema: tallyResponseSchema, id: 'RHSM tally' })
};

export { rhsmSchemas as default, rhsmSchemas };
