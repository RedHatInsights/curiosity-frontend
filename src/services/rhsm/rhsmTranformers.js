import {
  RHSM_API_RESPONSE_TALLY_DATA_TYPES as TALLY_DATA_TYPES,
  RHSM_API_RESPONSE_TALLY_META_TYPES as TALLY_META_TYPES,
  rhsmConstants
} from './rhsmConstants';

const rhsmTally = response => {
  const updatedResponse = {};
  const { [rhsmConstants.RHSM_API_RESPONSE_DATA]: data = [], [rhsmConstants.RHSM_API_RESPONSE_META]: meta = {} } =
    response || {};

  updatedResponse.data = data.map(
    (
      { [TALLY_DATA_TYPES.DATE]: date, [TALLY_DATA_TYPES.VALUE]: value, [TALLY_DATA_TYPES.HAS_DATA]: hasData },
      index
    ) => ({
      x: index,
      y: value,
      date,
      hasData
    })
  );

  updatedResponse.meta = {
    count: meta[TALLY_META_TYPES.COUNT],
    metricId: meta[TALLY_META_TYPES.METRIC_ID],
    productId: meta[TALLY_META_TYPES.PRODUCT],
    totalMonthlyDate: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.DATE],
    totalMonthlyHasData: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.HAS_DATA],
    totalMonthlyValue: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.VALUE]
  };

  return updatedResponse;
};

const rhsmTranformers = {
  tally: rhsmTally
};

export { rhsmTranformers as default, rhsmTranformers, rhsmTally };
