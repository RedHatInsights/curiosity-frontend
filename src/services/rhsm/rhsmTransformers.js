import moment from 'moment';
import {
  RHSM_API_RESPONSE_INSTANCES_DATA_TYPES as INSTANCES_DATA_TYPES,
  RHSM_API_RESPONSE_INSTANCES_META_TYPES as INSTANCES_META_TYPES,
  RHSM_API_RESPONSE_TALLY_DATA_TYPES as TALLY_DATA_TYPES,
  RHSM_API_RESPONSE_TALLY_META_TYPES as TALLY_META_TYPES,
  rhsmConstants
} from './rhsmConstants';
import { dateHelpers } from '../../common';

/**
 * FixMe: If RHSM Instances is deprecating Hosts we're missing a property, number_of_guests
 */
/**
 * Parse RHSM instances response for caching.
 *
 * @param {object} response
 * @returns {object}
 */
const rhsmInstances = response => {
  const updatedResponse = {};
  const { [rhsmConstants.RHSM_API_RESPONSE_DATA]: data = [], [rhsmConstants.RHSM_API_RESPONSE_META]: meta = {} } =
    response || {};
  const metaMeasurements = meta[INSTANCES_META_TYPES.MEASUREMENTS];

  updatedResponse.data = data.map(
    ({
      [INSTANCES_DATA_TYPES.MEASUREMENTS]: measurements,
      [INSTANCES_DATA_TYPES.SUBSCRIPTION_MANAGER_ID]: subscriptionManagerId,
      [INSTANCES_DATA_TYPES.NUMBER_OF_GUESTS]: numberOfGuests,
      ...dataResponse
    }) => {
      const updatedData = {
        numberOfGuests,
        subscriptionManagerId,
        ...dataResponse
      };

      metaMeasurements?.forEach((measurement, index) => {
        updatedData[measurement] = measurements[index];
      });

      return updatedData;
    }
  );

  updatedResponse.meta = {
    count: meta[INSTANCES_META_TYPES.COUNT],
    productId: meta[INSTANCES_META_TYPES.PRODUCT]
  };

  return updatedResponse;
};

/**
 * Parse RHSM tally response for caching.
 *
 * @param {object} response
 * @returns {object}
 */
const rhsmTally = response => {
  const updatedResponse = {};
  const { [rhsmConstants.RHSM_API_RESPONSE_DATA]: data = [], [rhsmConstants.RHSM_API_RESPONSE_META]: meta = {} } =
    response || {};
  const currentDay = moment.utc(dateHelpers.getCurrentDate()).format('MM-D-YYYY');

  updatedResponse.data = data.map(
    (
      { [TALLY_DATA_TYPES.DATE]: date, [TALLY_DATA_TYPES.VALUE]: value, [TALLY_DATA_TYPES.HAS_DATA]: hasData },
      index
    ) => ({
      x: index,
      y: value,
      date,
      hasData,
      isCurrentDate: moment.utc(date).format('MM-D-YYYY') === currentDay
    })
  );

  updatedResponse.meta = {
    count: meta[TALLY_META_TYPES.COUNT],
    cloudigradeHasMismatch: meta[TALLY_META_TYPES.HAS_CLOUDIGRADE_MISMATCH],
    metricId: meta[TALLY_META_TYPES.METRIC_ID],
    productId: meta[TALLY_META_TYPES.PRODUCT],
    totalMonthlyDate: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.DATE],
    totalMonthlyHasData: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.HAS_DATA],
    totalMonthlyValue: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.VALUE]
  };

  return updatedResponse;
};

const rhsmTransformers = {
  instances: rhsmInstances,
  tally: rhsmTally
};

export { rhsmTransformers as default, rhsmTransformers, rhsmInstances, rhsmTally };
