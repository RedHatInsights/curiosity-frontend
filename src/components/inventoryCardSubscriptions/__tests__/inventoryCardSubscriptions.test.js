import React from 'react';
import { InventoryCardSubscriptions } from '../inventoryCardSubscriptions';
import { RHSM_API_QUERY_SET_TYPES } from '../../../services/rhsm/rhsmConstants';

describe('InventoryCardSubscriptions Component', () => {
  it('should render a basic component', async () => {
    const props = {
      useProductInventoryConfig: () => ({ filters: [], settings: {} }),
      useProductInventoryQuery: () => ({
        [RHSM_API_QUERY_SET_TYPES.LIMIT]: 10,
        [RHSM_API_QUERY_SET_TYPES.OFFSET]: 0
      })
    };

    const component = await shallowComponent(<InventoryCardSubscriptions {...props} />);
    expect(component).toMatchSnapshot('basic render');
  });
});
