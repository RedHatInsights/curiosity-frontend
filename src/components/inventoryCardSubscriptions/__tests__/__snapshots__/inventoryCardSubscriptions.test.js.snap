// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryCardSubscriptions Component should render a basic component: basic render 1`] = `
<InventoryCard
  cardActions={null}
  isDisabled={false}
  perPageDefault={10}
  t={[Function]}
  useGetInventory={[Function]}
  useOnColumnSort={[Function]}
  useOnPage={[Function]}
  useProduct={[Function]}
  useProductInventoryConfig={[Function]}
  useProductInventoryQuery={[Function]}
  useSession={[Function]}
/>
`;
